Assets/Scripts/Layers/Adapters/ChainOfResponsibility.cs

```diff
@@ -0,0 +1,134 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.IO;
+using UnityEngine;
+using UnityEngine.Events;
+using UnityEngine.Networking;
+
+namespace Netherlands3D.Twin
+{
+    public class ChainOfResponsibility : MonoBehaviour
+    {
+        [Header("Data type adapters")] [Space(5)]
+        [SerializeField] private ScriptableObject[] dataTypeAdapters;
+        private IDataTypeAdapter[] dataTypeAdapterInterfaces;
+        public UnityEvent OnAdapterFound = new();
+
+        [Header("Events invoked on failures")] [Space(5)]
+        public UnityEvent<string> CouldNotFindAdapter = new();
+        public UnityEvent<string> OnDownloadFailed = new();
+
+        private string targetUrl = "";
+        public string TargetUrl { get => targetUrl; set => targetUrl = value; }
+
+        private Coroutine chain;
+
+        private void OnDisable() {
+            if(chain != null)
+                StopCoroutine(chain);
+        }
+
+        /// <summary>
+        /// Determine the type of data using chain of responsibility
+        /// </summary>
+        /// <param name="url">Url to file or service</param>
+        public void DetermineAdapter(string url)
+        {
+            TargetUrl = url;
+            DetermineAdapter();
+        }
+        public void DetermineAdapter()
+        {
+            if(chain != null)
+                StopCoroutine(chain);
+
+            chain = StartCoroutine(DownloadAndCheckSupport(TargetUrl));
+        }
+
+        private IEnumerator DownloadAndCheckSupport(string url)
```

Comment:
Wouldn't it be more efficient to only download where needed? And thus delegate the downloading and checking to the adapter instead of always fetching contents?

Reply:
The idea was to only download once to a local cached file, and use that as input for all adapters. 
This way only one download is done (most adapters need to check the content anyway to determine what kind of data it is)

Reply:
Seperate ticket for determining when file download/mimetype checks should happen to avoid early downloads (big files, slow etc.)